Samples/WinMLSamplesGallery/TransformAsync/SegmentModel.h

```diff
@@ -0,0 +1,152 @@
+#define USE_PIX
+//#define DBG
+#include <winrt/Microsoft.AI.MachineLearning.Experimental.h>
+#include <winrt/Microsoft.AI.MachineLearning.h>
+#include <Windows.AI.MachineLearning.native.h>
+#include <winrt/Windows.Graphics.DirectX.Direct3D11.h>
+#include <windows.media.core.interop.h>
+#include <winrt/Windows.Devices.Display.Core.h>
+#include <strsafe.h>
+#include <wtypes.h>
+#include <winrt/base.h>
+#include <dxgi.h>
+#include <d3d11.h>
+#include <mutex>
+#include <winrt/windows.foundation.collections.h>
+#include <winrt/Windows.Media.h>
+//#include <DXProgrammableCapture.h>
+#include "External/common.h"
+
+
+using namespace winrt::Microsoft::AI::MachineLearning;
+using namespace winrt::Microsoft::AI::MachineLearning::Experimental;
+using namespace winrt::Windows::Graphics::DirectX::Direct3D11;
+using namespace winrt::Windows::Media;
+
+// Model-agnostic helper LearningModels
+LearningModel Normalize0_1ThenZScore(long height, long width, long channels, const std::array<float, 3>& means, const std::array<float, 3>& stddev);
+LearningModel ReshapeFlatBufferToNCHW(long n, long c, long h, long w);
+LearningModel Invert(long n, long c, long h, long w);
+
+class StreamModelBase
+{
+public:
+	StreamModelBase(): 
+		m_inputVideoFrame(NULL),
+		m_outputVideoFrame(NULL),
+		m_session(NULL),
+		m_binding(NULL), 
+		m_syncStarted(FALSE)
+	{}
+	StreamModelBase(int w, int h) :
+		m_inputVideoFrame(NULL),
+		m_outputVideoFrame(NULL),
+		m_session(NULL),
+		m_binding(NULL),
+		m_syncStarted(FALSE)
+	{}
+	virtual ~StreamModelBase() {
+		if(m_session) m_session.Close();
+		if(m_binding) m_binding.Clear();
+		if (m_inputVideoFrame) m_inputVideoFrame.Close();
+		if (m_outputVideoFrame) m_outputVideoFrame.Close(); 
+	};
+
+	virtual void InitializeSession(int w, int h) =0;
+	virtual void Run(IDirect3DSurface src, IDirect3DSurface dest) =0;
+	
+	// Synchronous eval status
+	BOOL m_syncStarted; 
+	VideoFrame m_outputVideoFrame;
+	static const int m_scale = 5;
+
+protected:
+	// Cache input frames into a shareable d3d-backed VideoFrame
+	void SetVideoFrames(VideoFrame inVideoFrame, VideoFrame outVideoFrame) 
+	{
+		if (true || !m_videoFramesSet)
+		{
+			auto device = m_session.Device().Direct3D11Device();
+			auto inDesc = inVideoFrame.Direct3DSurface().Description();
+			auto outDesc = outVideoFrame.Direct3DSurface().Description();
+			/*
+				NOTE: VideoFrame::CreateAsDirect3D11SurfaceBacked takes arguments in (width, height) order
+				whereas every model created with LearningModelBuilder takes arguments in (height, width) order. 
```

Comment:
>whereas every model created with LearningModelBuilder takes arguments in (height, width) order.  [](http://example.com/codeflow?start=4&length=96)

Yeah, it's pretty weird. Everything I can think of in the world lists x before y, width before height, *except* ML:

8x11 sheet of paper - 8 wide, 11 tall
16x18 frame - 16 wide, 18 tall
640x480 monitor - meaning 640 wide, 480 tall
32x30 jeans - meaning 32 inches wide, 30 tall
2D point coordinates - x then y

Reply:
see my other point about leraningmodelbuilders being hard to debug 😅